### PR TITLE
Copy and move constructor added in GRU layer.

### PR DESCRIPTION
**Pull request by AvikantSrivastava**
In reference to https://github.com/mlpack/mlpack/issues/2625
In this PR I have added definition of move and copy constructor to the GRU layer.

### CODE REVIEW

**Review by zoq — COMMENTED** [review]
Here are the two test files that are relevant for GRU:

- https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/ann_layer_test.cpp
- https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/recurrent_network_test.cpp

You can put the copy/move test into the `ann_layer_test.cpp` file.

**Review by zoq — COMMENTED** [review]
That's one example test for copy/move, you could use the same test for the GRU layer:

- https://github.com/mlpack/mlpack/pull/2711/files#diff-338dcffcd150ae0b9b5c3c700d8aac8dbe9bd125a3f411d4fbccc2dd99a41be6

**Comment by AvikantSrivastava** [comment]
> That's one example test for copy/move, you could use the same test for the GRU layer:
> 
> * https://github.com/mlpack/mlpack/pull/2711/files#diff-338dcffcd150ae0b9b5c3c700d8aac8dbe9bd125a3f411d4fbccc2dd99a41be6

Thanks, I'll look into it

**Comment by AvikantSrivastava** [comment]
https://github.com/mlpack/mlpack/pull/2723/commits/0ca29e9718c38ef07dbe29da02efb3a9d4da6e44
https://github.com/mlpack/mlpack/pull/2723/commits/7584a0e794e180d89e8a5ff1439afd6ffc0f2ffe
@zoq 
Code formatting problem fixed, let me know if the current format is write

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 1]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+// Copy Constructor definition
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU(
+  const GRU& layer) : inSize(layer.inSize),
+  outSize(layer.outSize),
+  rho(layer.rho),
+  weights(layer.weights)
+
+{
+  // Nothing to do here
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:88]:
```suggestion
  // Nothing to do here.
```

Pedantic style issue, missing stop at the end.

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 2]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+// Move Constructor definition
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU(
+  GRU&& layer) : inSize(std::move(layer.inSize)),
+  outSize(std::move(layer.outSize)),
+  rho(std::move(layer.rho)),
+  weights(std::move(layer.weights))
+
+{
+  // Nothing to do here
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:100]:
```suggestion
  // Nothing to do here.
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 3]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
   gradIterator = outParameter.end();
 }
 
+// Copy Constructor definition
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU(
+  const GRU& layer) : inSize(layer.inSize),
+  outSize(layer.outSize),
+  rho(layer.rho),
+  weights(layer.weights)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:85]:
```suggestion
GRU<InputDataType, OutputDataType>::GRU(const GRU& layer) :
    inSize(layer.inSize),
    outSize(layer.outSize),
    rho(layer.rho),
    weights(layer.weights)
```
I tend to be very picky about the style :)

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 4]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:86]:
```suggestion
```

Remove the extra line here.

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 5]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+}
+
+// Move Constructor definition
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU(
+  GRU&& layer) : inSize(std::move(layer.inSize)),
+  outSize(std::move(layer.outSize)),
+  rho(std::move(layer.rho)),
+  weights(std::move(layer.weights))
+
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:98]:
```suggestion
GRU<InputDataType, OutputDataType>::GRU(GRU&& layer) :
    inSize(std::move(layer.inSize)),
    outSize(std::move(layer.outSize)),
    rho(std::move(layer.rho)),
    weights(std::move(layer.weights))
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 6]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+{
+  // Nothing to do here
+}
+
+// Copy operator definition
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>&
+  GRU<InputDataType, OutputDataType>::
+  operator=(const GRU& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:107]:
```suggestion
  GRU<InputDataType, OutputDataType>::operator=(const GRU& layer)
```

Looks like this fits on the same line.

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 7]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+  GRU&& layer) : inSize(std::move(layer.inSize)),
+  outSize(std::move(layer.outSize)),
+  rho(std::move(layer.rho)),
+  weights(std::move(layer.weights))
+
+{
+  // Nothing to do here
+}
+
+// Copy operator definition
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:103]:
```suggestion
// Copy operator definition.
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 8]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+  const GRU& layer) : inSize(layer.inSize),
+  outSize(layer.outSize),
+  rho(layer.rho),
+  weights(layer.weights)
+
+{
+  // Nothing to do here
+}
+
+// Move Constructor definition
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:91]:
```suggestion
// Move Constructor definition
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 9]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
   gradIterator = outParameter.end();
 }
 
+// Copy Constructor definition
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:79]:
```suggestion
// Copy Constructor definition.
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 10]
```diff
@@ -72,6 +72,19 @@ class GRU
...
+  // Copy Constructor
+  GRU(const GRU &layer);
+
+  // Move Constructor
+  GRU(GRU &&);
+
+  // Copy assignment operator
+  GRU &operator=(const GRU &layer);
+
+  // Move assignment operator
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:84]:
```suggestion
  // Move assignment operator.
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 11]
```diff
@@ -72,6 +72,19 @@ class GRU
       const size_t outSize,
       const size_t rho = std::numeric_limits<size_t>::max());
 
+  // Copy Constructor
+  GRU(const GRU &layer);
+
+  // Move Constructor
+  GRU(GRU &&);
+
+  // Copy assignment operator
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:81]:
```suggestion
  // Copy assignment operator.
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 12]
```diff
@@ -72,6 +72,19 @@ class GRU
       const size_t outSize,
       const size_t rho = std::numeric_limits<size_t>::max());
 
+  // Copy Constructor
+  GRU(const GRU &layer);
+
+  // Move Constructor
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:78]:
```suggestion
  // Move constructor.
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 13]
```diff
@@ -72,6 +72,19 @@ class GRU
       const size_t outSize,
       const size_t rho = std::numeric_limits<size_t>::max());
 
+  // Copy Constructor
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:75]:
```suggestion
  // Copy constructor.
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 14]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+    inSize = layer.inSize;
+    outSize = layer.outSize;
+    rho = layer.rho;
+    weights = layer.weights;
+  }
+  return *this;
+}
+
+// Move operator definition
+
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:120]:
```suggestion
// Move operator definition.
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 15]
```diff
@@ -76,6 +76,64 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>&
+  GRU<InputDataType, OutputDataType>::
+  operator=(GRU&& layer)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:124]:
```suggestion
  GRU<InputDataType, OutputDataType>::operator=(GRU&& layer)
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 16]
```diff
@@ -72,6 +72,19 @@ class GRU
...
+  GRU(const GRU &layer);
+
+  // Move Constructor
+  GRU(GRU &&);
+
+  // Copy assignment operator
+  GRU &operator=(const GRU &layer);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:82]:
```suggestion
  GRU &operator=(const GRU& layer);
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 17]
```diff
@@ -72,6 +72,19 @@ class GRU
       const size_t outSize,
       const size_t rho = std::numeric_limits<size_t>::max());
 
+  // Copy Constructor
+  GRU(const GRU &layer);
+
+  // Move Constructor
+  GRU(GRU &&);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:79]:
```suggestion
  GRU(GRU&&);
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 18]
```diff
@@ -72,6 +72,19 @@ class GRU
       const size_t outSize,
       const size_t rho = std::numeric_limits<size_t>::max());
 
+  // Copy Constructor
+  GRU(const GRU &layer);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:76]:
```suggestion
  GRU(const GRU& layer);
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 19]
```diff
@@ -72,6 +72,19 @@ class GRU
...
+
+  // Move Constructor
+  GRU(GRU &&);
+
+  // Copy assignment operator
+  GRU &operator=(const GRU &layer);
+
+  // Move assignment operator
+  GRU &operator=(GRU &&layer);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/gru.hpp:85]:
```suggestion
  GRU &operator=(GRU&& layer);
```

**Review by zoq — COMMENTED** [review]
All of the comments are really minor, I just like to see we use the same coding style across the codebase.

**Comment by AvikantSrivastava** [comment]
@zoq Changes noted and committed to the PR.

**Comment by AvikantSrivastava** [comment]
@zoq I have added tests for GRU copy and move constructor in https://github.com/mlpack/mlpack/pull/2723/commits/294b43c4b5ca699158e427b525ac4672692cafdd
In this commit, I did not compile the layer into a model, simply checked copying and moving operators.

Let me know how to proceed next. :)

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 20]
```diff
@@ -76,6 +76,59 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+
+// Move operator definition.
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>&
+  GRU<InputDataType, OutputDataType>::operator=(GRU&& layer)
+{
+  if (this != &layer)
+  {
+    inSize = layer.inSize;
+    outSize = layer.outSize;
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:124]:
```suggestion
    outSize = std::move(layer.outSize);
```
Same as here

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 21]
```diff
@@ -76,6 +76,59 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+}
+
+// Move operator definition.
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>&
+  GRU<InputDataType, OutputDataType>::operator=(GRU&& layer)
+{
+  if (this != &layer)
+  {
+    inSize = layer.inSize;
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:123]:
```suggestion
    inSize = std::move(layer.inSize);
```
I am not sure if we need to use move here, but I do not see why not

File: src/mlpack/tests/ann_layer_test.cpp [hunk 22]
```diff
@@ -1484,6 +1484,64 @@ TEST_CASE("ForwardGRULayerTest", "[ANNLayerTest]")
...
+  // Provide input of all ones.
+  arma::mat input1 = arma::ones(3, 1);
+
+  // Declaring two ouput matrices for each layer.
+  arma::mat output1;
+  arma::mat output2;
+
+  // Forward pass through layer1.
+  layer1.Forward(input, output1);
+  layer2(std::move(*layer1));
```
Comment (himanshupathak21061998) [inline src/mlpack/tests/ann_layer_test.cpp:1534]: I think it `layer2` should be initialized here.
```
GRU<> layer2(std::move(*layer1));
```

**Comment by AvikantSrivastava** [comment]
@himanshupathak21061998 @shrit @zoq 
I did not use use `CheckCopyFunction()` and `CheckMoveFunction()` in [CheckCopyGRUTest](https://github.com/mlpack/mlpack/pull/2723/files#diff-b05ef9758cf62593497c2fcfd87ff1d915d1b4ca8147d999480e7e07ad92fb6aR1490) and [CheckMovingGRUTest](https://github.com/mlpack/mlpack/pull/2723/files#diff-b05ef9758cf62593497c2fcfd87ff1d915d1b4ca8147d999480e7e07ad92fb6aR1519)

Should I write functions as written in https://github.com/mlpack/mlpack/pull/2583 ?

**Comment by Abilityguy** [comment]
Hi @AvikantSrivastava , PR #2745 adds ```CheckRNNCopyFunction``` and ```CheckRNNMoveFunction``` testing wrappers. I think you can use those to write GRU copy and move tests by adding the layer in a model. This will also help to maintain a testing style similar to the ones used in ```FFN``` tests.

**Comment by AvikantSrivastava** [comment]
@zoq @shrit @himanshupathak21061998 I am getting a segmentation fault error here https://github.com/mlpack/mlpack/pull/2723/files#diff-b05ef9758cf62593497c2fcfd87ff1d915d1b4ca8147d999480e7e07ad92fb6aR1567

**Comment by shrit** [comment]
@AvikantSrivastava Can you add more output from the segmentation fault

**Comment by AvikantSrivastava** [comment]
> @AvikantSrivastava Can you add more output from the segmentation fault?

The log in the CI build is pointing to the very first line 
https://dev.azure.com/mlpack/mlpack/_build/results?buildId=5261&view=logs&j=24d3abe3-ef0b-5deb-3aab-64d839de2c3c&t=b3d23cc5-b695-5043-0f03-2084bf2ff0b5&l=34

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 23]
```diff
@@ -76,6 +76,59 @@ GRU<InputDataType, OutputDataType>::GRU(
   gradIterator = outParameter.end();
 }
 
+// Copy Constructor definition.
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU(const GRU& layer) :
+    inSize(layer.inSize),
+    outSize(layer.outSize),
+    rho(layer.rho),
+    weights(layer.weights)
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:85]: Could you try including batch_size, forward_step, backward_step, gradient_step and other parameters in the initialisation list ? I believe that should solve the segmentation fault.

Comment (AvikantSrivastava) [reply src/mlpack/methods/ann/layer/gru_impl.hpp]: Thanks @Abilityguy 
I have made the changes that you have suggested
Hopefully, the code does not throw the same error this time

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 24]
```diff
@@ -76,6 +76,83 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU(GRU&& layer) :
+    inSize(std::move(layer.inSize)),
+    outSize(std::move(layer.outSize)),
+    rho(std::move(layer.rho)),
+    batchSize(std::move(layer.batchSize)),
+    forwardStep(std::move(layer.forwardStep)),
+    backwardStep(std::move(layer.backwardStep)),
+    gradientStep(std::move(layer.gradientStep)),
+    weights(std::move(layer.weights)),
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:106]:
```suggestion
    weights(std::move(layer.weights)),
    outParameter(std::move(layer.outParameter)),
    allZeros(std::move(layer.allZeros)),
    prevOutput(std::move(layer.prevOutput)),
    backIterator(std::move(layer.backIterator)),
    gradIterator(std::move(layer.gradIterator)),
    prevError(std::move(layer.prevError)),
    input2GateModule(std::move(layer.input2GateModule)),
    output2GateModule(std::move(layer.output2GateModule)),
    outputHidden2GateModule(std::move(layer.outputHidden2GateModule)),
    inputGateModule(std::move(layer.inputGateModule)),
    hiddenStateModule(std::move(layer.hiddenStateModule)),
    forgetGateModule(std::move(layer.forgetGateModule)),
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 25]
```diff
@@ -76,6 +76,83 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+  if (this != &layer)
+  {
+    inSize = layer.inSize;
+    outSize = layer.outSize;
+    rho = layer.rho;
+    batchSize = layer.batchSize;
+    forwardStep = layer.forwardStep;
+    backwardStep = layer.backwardStep;
+    gradientStep = layer.gradientStep;
+    weights = layer.weights;
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:127]:
```suggestion
    weights = layer.weights;
    outParameter = layer.outParameter;
    allZeros = layer.allZeros;
    prevOutput = layer.prevOutput;
    backIterator = layer.backIterator;
    gradIterator = layer.gradIterator;
    prevError = layer.prevError;
    input2GateModule = layer.input2GateModule;
    output2GateModule = layer.output2GateModule;
    outputHidden2GateModule = layer.outputHidden2GateModule;
    inputGateModule = layer.inputGateModule;
    hiddenStateModule = layer.hiddenStateModule;
    forgetGateModule = layer.forgetGateModule;
```

File: src/mlpack/methods/ann/layer/gru_impl.hpp [hunk 26]
```diff
@@ -76,6 +76,83 @@ GRU<InputDataType, OutputDataType>::GRU(
...
+  if (this != &layer)
+  {
+    inSize = std::move(layer.inSize);
+    outSize = std::move(layer.outSize);
+    rho = std::move(layer.rho);
+    batchSize = std::move(layer.batchSize);
+    forwardStep = std::move(layer.forwardStep);
+    backwardStep = std::move(layer.backwardStep);
+    gradientStep = std::move(layer.gradientStep);
+    weights = std::move(layer.weights);
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/gru_impl.hpp:148]:
```suggestion
    weights = std::move(layer.weights);
    outParameter = std::move(layer.outParamater);
    allZeros = std::move(layer.allZeros);
    prevOutput = std::move(layer.prevOutput);
    backIterator = std::move(layer.backIterator);
    gradIterator = std::move(layer.gradIterator);
    prevError = std::move(layer.prevError);
    input2GateModule = std::move(layer.input2GateModule);
    output2GateModule = std::move(layer.output2GateModule);
    outputHidden2GateModule = std::move(layer.outputHidden2GateModule);
    inputGateModule = std::move(layer.inputGateModule);
    hiddenStateModule = std::move(layer.hiddenStateModule);
    forgetGateModule = std::move(layer.forgetGateModule);
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 27]
```diff
@@ -72,6 +72,19 @@ class GRU
       const size_t outSize,
       const size_t rho = std::numeric_limits<size_t>::max());
 
+  // Copy constructor.
+  GRU(const GRU& layer);
+
+  // Move constructor.
+  GRU(GRU&&);
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/gru.hpp:79]:
```suggestion
  GRU(GRU&& layer);
```

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 28]
```diff
@@ -72,6 +72,19 @@ class GRU
...
+
+  // Copy assignment operator.
+  GRU &operator=(const GRU& layer);
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/gru.hpp:82]:
```suggestion
  GRU& operator=(const GRU& layer);
```
To match the style in other files.

File: src/mlpack/methods/ann/layer/gru.hpp [hunk 29]
```diff
@@ -72,6 +72,19 @@ class GRU
...
+
+  // Move assignment operator.
+  GRU &operator=(GRU&& layer);
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/gru.hpp:85]:
```suggestion
  GRU& operator=(GRU&& layer);
```

**Review by Abilityguy — CHANGES_REQUESTED** [review]
Hi @AvikantSrivastava ,
I went through your code.
I made some changes and the tests were passing on my local system.
I have suggested those changes. Hopefully we should get a green build now.

File: src/mlpack/methods/ann/layer/gru.hpp (hunk 29 above)
Comment (AvikantSrivastava) [reply src/mlpack/methods/ann/layer/gru.hpp]: @Abilityguy Thank you so much for reviewing the PR, I was not active for some time and the PR went stale.
I have made the changes that you suggested, and I will resolve the bugs and complete the PR.
Thanks again :)

**Comment by AvikantSrivastava** [comment]
@Abilityguy  Thank you so much for reviewing the PR, I was not active for some time and the PR went stale.
I have made the changes that you suggested, and I will resolve the bugs and complete the PR.
Thanks again :)

**Comment by rcurtin** [comment]
Hey @AvikantSrivastava, if you'd like to keep this one moving, do you think you could debug the failing tests?  I'd be happy to see this get merged, but I'd prefer to wait to review it until all the tests are passing. :+1:  Let me know if I can clarify any of the errors or anything.

**Comment by AvikantSrivastava** [comment]
@rcurtin I would love to resolve the problem here, but I'm not able to follow the bugs here, Can you help me solve it.

**Comment by rcurtin** [comment]
I think that the primary issue is that when we copy a `LayerTypes<>`, we actually have to use a `CopyVisitor`, since `LayerTypes<>` actually holds a pointer internally.

For instance, take a look at the copy constructor of the `Sequential` and `Recurrent` layers.  The patterns used there for the internally-held `network` will need to be used here for `network`, `input2GateModule`, and all the other members of `GRU` that are (or hold) a `LayerTypes<>`. :+1:

**Pull request closed without merging**